Sources/SotoCore/Credential/ConfigFileCredentialProvider.swift

```diff
@@ -12,144 +12,91 @@
+        from credentialsFilePath: String,
+        configFilePath: String,
+        for profile: String,
+        context: CredentialProviderFactory.Context
     ) -> EventLoopFuture<CredentialProvider> {
-        let filePath = Self.expandTildeInFilePath(credentialsFilePath)
-
-        return fileIO.openFile(path: filePath, eventLoop: eventLoop)
-            .flatMap { handle, region in
-                fileIO.read(fileRegion: region, allocator: ByteBufferAllocator(), eventLoop: eventLoop).map { ($0, handle) }
-            }
-            .flatMapThrowing { byteBuffer, handle in
-                try handle.close()
-                return try Self.sharedCredentials(from: byteBuffer, for: profile)
-            }
-    }
-
-    static func sharedCredentials(from byteBuffer: ByteBuffer, for profile: String) throws -> StaticCredential {
-        let string = byteBuffer.getString(at: 0, length: byteBuffer.readableBytes)!
-        var parser: INIParser
-        do {
-            parser = try INIParser(string)
-        } catch INIParser.Error.invalidSyntax {
-            throw ConfigFileError.invalidCredentialFileSyntax
-        }
-
-        guard let config = parser.sections[profile] else {
-            throw ConfigFileError.missingProfile(profile)
+        return ConfigFileLoader.loadSharedCredentials(
+            credentialsFilePath: credentialsFilePath,
+            configFilePath: configFilePath,
+            profile: profile,
+            context: context
+        )
+        .flatMapThrowing { credentials, config in
+            return try credentialProvider(from: credentials, config: config, context: context)
         }
+    }
 
-        guard let accessKeyId = config["aws_access_key_id"] else {
-            throw ConfigFileError.missingAccessKeyId
+    /// Generate credential provider based on shared credentials and profile configuration
+    ///
+    /// Credentials file settings have precedence over profile configuration settings
+    /// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence
+    ///
+    /// - Parameters:
+    ///   - credentialsBuffer: contents of AWS shared credentials file (usually `~/.aws/credentials`
+    ///   - configByteBuffer: contents of AWS profile configuration file (usually `~/.aws/config`
+    ///   - context: credential provider factory context
+    /// - Returns: Credential Provider (StaticCredentials or STSAssumeRole)
+    static func credentialProvider(
+        from credentials: ConfigFileLoader.ProfileCredentials,
+        config: ConfigFileLoader.ProfileConfig?,
+        context: CredentialProviderFactory.Context
+    ) throws -> CredentialProvider {
+        // If `role_arn` and `sourcer_profile` are defined, temporary credentials must be loaded via STS Assume Role operation
+        if let roleArn = credentials.roleArn ?? config?.roleArn,
+           let _ = credentials.sourceProfile ?? config?.sourceProfile
+        {
```

Comment:
Not sure I understand your question. 

`ConfigFileLoader` loads `credentials` and `config`. In the case of `credentials`, it will load `accessKey` and `secretAccessKey` if a `sourceProfile` is present (see `ConfigFileParser.parseCredentials`).

Reply:
Returning a `role_arn` and `credential_source` without credentials is a valid response. You could structure your code as follows
```swift
if role_arn {
    if credentials {
        STSAssumeRoleWithStaticCredentials()
    } else if credentialSource {
        STSAssumeRoleWIthOneOfEC2, ECS, Environment()
    } else {
        Error()
    }
} else if credentials {
    StaticCredential()
} else {
    Error()
}
```

Reply:
Actually, `credentials.accessKey` and `credentials.secretAccessKey` are non-optional, so they can never be null at this point. `ConfigFileLoader` will throw an exception if they were missing. My understanding is that credentials are also needed for `credential_source`, but I might be wrong. If that is the case, we can update `ConfigFileLoader` as needed once we add support for `credential_source` (preferably on a separate PR).

While the code structure you suggested above is good, I opted for `if/return` to avoid nesting multiple `if/else`. I feel the code reads better this way.

STS AssumeRole operation should be made only if `sourceProfile` is defined. I think relying only on the existence of `role_arn` and credentials could be misleading, making it hard to understand what is happening.  To make this process clear, I'm going to move this configuration logic to `ConfigFileLoader`, and use a `switch` here instead to determine which provider to create.

Reply:
Ok, looks like `credential_source` does not require access/secret keys, so I've updated the code accordingly.

Reply:
The enum is a good call, you just need to implement `credential_source` and I think we're done. 

Do you know what you need to do here? Set the credential provider in `STSAssumeRoleCredentialProvider.init` to `.environment`, `.ecs` or `.ec2`, based on the value of `credential_source`.

Reply:
I don't have a way set up to test this functionality, so I'd like to leave that implementation for another PR.